### PR TITLE
#344 update DetailPage

### LINKED ISSUE #344: Product details heading doesn't make sense
## Expected Behaviour

When drilling down in the web UI to view a product, the descriptions on the page should be clear and easy to understand.

## Current Behaviour

Example header for Amoxicillin Oral Capsule 250mg:
image

## Possible Solution

If possible I think it would be good to display the product hierarchy, something like Amoxicillin / Oral / Capsule / 250mg

## Steps to Reproduce (bugs only, delete if not applicable)

1. Open UC web (https://codes.msupply.foundation/) and search for a drug e.g. Amoxicillin
2. Click on a Drug 
3. Select 'Forms' and expand the tree as far as possible
4. Select a product at the lowest level
5. Inspect UI

### PR DESCRIPTION
**Pull request by wlthomson**
Fixes #344, #353, #361, #366.

## Description

Updates `DetailPage` to support recursive traversal of entities and properties. 

- Updates `DetailPage` to be more recursive, supporting nested entities and properties.
- Adds entity grouping by `type`, supporting entities with children of different types.
- Adds support for copying entity codes.
- Adds name preprocessing to `data-loader` (nested items and headers now make a lot more sense!). 
- Retains support for accessing entities at any level via editing the url.

### Documentation:

Tick below if one of the following applies:

1. Documentation has been added to cover changes introduced by this PR.
2. This PR requires no changes to documentation.

- [x] Documentation OK.

### Tests:

Tick below if one of the following applies:

1. Tests have been added to cover changes introduced by this PR.
2. This PR requires no new tests.

- [x] Tests OK.

### CODE REVIEW

**Review by katherine-sussol — APPROVED** [review]
Looks good to me, so much easier to use now! 

Some additional react validation gripes that I didn't get to the bottom of but still working even with them so 🤷 :
image

File: unified-codes/apps/web/src/components/containers/detail/DetailProduct.tsx [hunk 1]
```diff
@@ -1,88 +0,0 @@
-import * as React from 'react';
```
Comment (katherine-sussol) [inline unified-codes/apps/web/src/components/containers/detail/DetailProduct.tsx:1]: 👋  Bye-bye - Going to miss you, but not going to miss all the confused end-users!

File: unified-codes/apps/web/src/components/containers/detail/DetailPropertyList.tsx [hunk 2]
```diff
@@ -1,50 +1,96 @@
...
+          <PropertyListToggleItemIcon />
+        </ListItemIcon>
+        <PropertyListToggleItemText />
+      </ListItem>
+    ) : null;
 
-  if (!propertyCount) return null;
+  const PropertyListChildItems = React.useCallback(() => {
+    const childItems = properties?.map((property: IProperty) => {
+      return <DetailPropertyListItem property={property} />;
```
[the remark this answers is not in the document]
Comment (wlthomson) [reply unified-codes/apps/web/src/components/containers/detail/DetailPropertyList.tsx]: 5e1638331f21b9c45e718d06ba8e0f3dddc6c751.

**Comment by wlthomson** [comment]
Hopefully working ok now!

Still quite messy, but I think this PR already pretty big. Cleanup can be done in dedicated issue?

Some things still to be done:

- Extract components for `DetailEntityTypeList` and `DetailEntityTypeListeItem`.
- Remove root level `entityFormats`, `propertyFormats` (better done with selectors, I think?).
- Tidy up styles + add component style typings.
- General tidyup (better variable names, comments, reducing duplicate code etc.)

**Pull request merged**